app/adapters/whats_app_adapter/three_sixty_dialog_inbound.rb

```diff
@@ -47,7 +52,35 @@ def trigger(event, *args)
       @callbacks[event].call(*args)
     end
 
-    def initialize_sender(whats_app_message)
+    def initialize_text
+      message = whats_app_message[:messages].first
+      message[:text]&.dig(:body) || message[:button]&.dig(:text) || supported_file(message)&.dig(:caption)
+    end
+
+    def ephemeral_data?
+      return if text.blank?
+
+      request_for_more_info? || unsubscribe_text? || resubscribe_text? || request_to_receive_message?
+    end
+
+    def trigger_appropriate_callback
+      callback =
+        if request_for_more_info?
+          REQUEST_FOR_MORE_INFO
+        elsif unsubscribe_text?
+          UNSUBSCRIBE_CONTRIBUTOR
+        elsif resubscribe_text?
+          RESUBSCRIBE_CONTRIBUTOR
+        elsif request_to_receive_message?
+          REQUEST_TO_RECEIVE_MESSAGE
+        end
+      params_array = [sender]
+      params_array.append(whats_app_message[:messages].first) if request_to_receive_message?
+
+      trigger(callback, *params_array)
```

Comment:
```suggestion
      if request_for_more_info?
        trigger(REQUEST_FOR_MORE_INFO, sender)
      elsif unsubscribe_text?
        trigger(UNSUBSCRIBE_CONTRIBUTOR, sender)
      elsif resubscribe_text?
        trigger(RESUBSCRIBE_CONTRIBUTOR, sender)
      elsif request_to_receive_message?
        trigger(REQUEST_TO_RECEIVE_MESSAGE, sender, whats_app_message[:message])
      end
```
I'm leaning towards "request changes" on this refactoring. The original implementation should work as well, though, just more code.

Reply:
I wrote it that way originally, but it involves more code duplication. All branches trigger the callback and all pass in the sender as the 2nd argument. The only difference is that the request to receive message passes in the first message as a third parameter.
I considered refactoring so they all have the same API and let the method handle extracting the data it's interested in.

Reply:
OK, so you prefer it as it is? I would suggest to put line 78 as close as possible to the call of `REQUEST_TO_RECEIVE_MESSAGE`. Otherwise it's a potential for bugs in my opinion.

I can hit "Approve" if you want.